ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
+        auto arg0_rank = arg0_shape.rank().get_length();
+        auto arg1_rank = arg1_shape.rank().get_length();
 
-        size_t arg0_rank = arg0_shape.size();
-        size_t arg1_rank = arg1_shape.size();
+        NGRAPH_CHECK((arg0_rank != 0 && arg1_rank != 0),
+                     "Scalars are not supported as MatMul inputs.");
 
+        // Temporary Dimension vectors to calculate output shape
+        std::vector<Dimension> arg0_shape_tmp(arg0_shape);
+        std::vector<Dimension> arg1_shape_tmp(arg1_shape);
+
+        // 1. Applying transpositions specified by optional `transpose_a` and `transpose_b`
+        // Only two right-most dimensions are swapped, other dimensions remain the same.
+        // Transpose attributes are ignored for 1D tensors.
         if (transpose_a && arg0_rank > 1)
         {
-            swap(arg0_shape_update[arg0_rank - 2], arg0_shape_update[arg0_rank - 1]);
+            swap(arg0_shape_tmp[arg0_rank - 2], arg0_shape_tmp[arg0_rank - 1]);
         }
         if (transpose_b && arg1_rank > 1)
         {
-            swap(arg1_shape_update[arg1_rank - 2], arg1_shape_update[arg1_rank - 1]);
+            swap(arg1_shape_tmp[arg1_rank - 2], arg1_shape_tmp[arg1_rank - 1]);
         }
 
-        if (arg0_rank == 1 && arg1_rank == 1)
+        // 2. One-dimensional tensors unsqueezing is applied to each input independently.
+        if (arg0_rank == 1)
         {
-            NGRAPH_CHECK(arg0_shape_update == arg1_shape_update, "Incompatible arg shapes");
-            output_shape = Shape{};
+            // If the first input is 1D tensor, it is unsqueezed to 2D tensor (row vector)
+            // by adding axes with size 1 at ROW_INDEX_DIM, to the left of the shape.
+            // For example {S} will be reshaped to {1, S}.
+            arg0_shape_tmp.insert(arg0_shape_tmp.begin(), 1);
+            arg0_rank = arg0_shape_tmp.size();
         }
-        else if (arg0_rank == 1)
+        if (arg1_rank == 1)
         {
-            // i.e., arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
-            NGRAPH_CHECK(arg0_shape_update[0] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
-            arg1_shape_update.erase(arg1_shape_update.begin() + arg1_rank - 2);
-            output_shape = arg1_shape_update;
+            // If the second input is 1D tensor, it is unsqueezed to 2D tensor (column vector)
+            // by adding axes with size 1 at COL_INDEX_DIM, to the right of the shape.
+            // For example {S} will be reshaped to {S, 1}.
+            arg1_shape_tmp.insert(arg1_shape_tmp.end(), 1);
+            arg1_rank = arg1_shape_tmp.size();
         }
-        else if (arg1_rank == 1)
+
+        // Check matrices dimensions compatibility,
+        // COL_INDEX_DIM of the first matrix has to match ROW_INDEX_DIM of the second matrix.
+        auto merged_dimension = Dimension::dynamic();
+        auto arg0_col_dim = arg0_shape_tmp[arg0_rank - 1];
```

Comment:
What happens if rank is dynamic?

Reply:
This validation is for static rank. For dynamic rank, `PartialShape::dynamic()` will be returned. https://github.com/openvinotoolkit/openvino/blob/e0d0ee561b9c0dc6668f08112d2e8924618d47f7/ngraph/core/src/op/matmul.cpp#L284-L299